### PR TITLE
Initial commit of ugrid loading code

### PR DESCRIPTION
**Pull request by alastair-gemmell**
Initial commit of ugrid_cf_reader.py

### CODE REVIEW

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Thanks @alastair-gemmell!

Two things we definitely need before merging:

- [ ] Any tests pertaining to `ugrid_cf_reader`
  - As discussed offline, we believe these tests are limited to [iris.tests.integration.ugrid.test_ugrid_load](https://github.com/SciTools/iris/blob/ng-vat/lib/iris/tests/integration/ugrid/test_ugrid_load.py), plus the accompanying [\_\_init__](https://github.com/SciTools/iris/blob/ng-vat/lib/iris/tests/integration/ugrid/__init__.py).
  - Will need access to iris-test-data - specific commit hash - see #7.
- [x] A `travis.yml` - see #7.
If some of #7 is not applicable at this point we can just implement what can and leave that issue open.

File: iris-ugrid/ugrid_cf_reader.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                mesh_location=mesh_location,
+                topology_dimension=topology_dimension,
+                node_coordinates=sorted(node_coordinates),
+            )
+        else:
+            # Add an empty 'cube.ugrid' to all cubes otherwise.
+            cube.ugrid = None
+        return
+
+    def __del__(self):
```
Comment (trexfeathers) [inline iris-ugrid/ugrid_cf_reader.py:229]: Can we rely on the `__del__` method just being in `CFReader`, thereby not needing the same method in `UGridCFReader`? Can test by checking that running `__del__` on the `UGridCFReader().cfreader` actually closes the dataset, and also by seeing if [iris.tests.integration.test_climatology](https://github.com/SciTools/iris/blob/master/lib/iris/tests/integration/test_climatology.py) still runs (known to have trouble if there is an open dataset).

File: iris-ugrid/ugrid_cf_reader.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        for meshname in find_mesh_names(self.dataset):
+            mesh = UGrid()
+            load_grid_from_nc_dataset(dataset, mesh, mesh_name=meshname)
+            meshes[meshname] = mesh
+        self.meshes = meshes
+
+        # Generate list of excluded variable names.
+        exclude_vars = list(meshes.keys())
+
+        temp_xios_fix = kwargs.pop("temp_xios_fix", False)
```
Comment (trexfeathers) [inline iris-ugrid/ugrid_cf_reader.py:128]: As discussed, and written up in #5, `temp_xios_false` is not planned to be used this time round. This code block should be reduced to just the standard handling (when `temp-xios-fix=False`.

**Comment by pp-mo** [comment]
Tiny note : I'm slightly wary of using hyphens in folder names, so I'm not keen on "iris-ugrid" as a folder name.

And in this case, it needs to acts as a package name, so I think it is a bad idea.
I **_personally_** also prefer to have a top-level folder that only contains the importable package, e.g. the "iris_ugrid  package" would be at `lib/iris_ugrid.py` or `lib/iris_ugrid_/__init__.py'.
This is because in a developer install, you need a source folder that you can add to PYTHONPATH.  If that is the repo top-level, then any config or build source become importable, e.e.g "import setup", "import config", which is generally undesirable.

I must say though, that not everyone thinks that is so important, and the use of 'lib' folders seems to be a bit old-school.
  * iris has `lib/iris/__init__.py`
  * cartopy has  `lib/cartopy/__init__.py`
  * iris-grib has `iris_grib/__init__.py`
  * tephi has 'tephi/__init__.py`

**Comment by alastair-gemmell** [comment]
Thanks @pp-mo!

Yes, changing the name to use an underscore sounds sensible -- iris_ugrid

WRT structure - I tried to think of which python libraries are most popular in our area and crib from them so I looked into how Dask, Zarr, Numpy and Xarray do it and they all have the same approach:

`dask/__init__.py`
`zarr/__init__.py`
`numpy/__init__.py`
`xarray/__init__.py`

That was the reasoning behind the structure I chose, but happy to change if you and others think it would be better to go with the lib option.

**Comment by alastair-gemmell** [comment]
Still a WIP on the required changes but now updated e.g. with renamed iris-ugrid -> iris_ugrid

File: iris_ugrid/tests/integration/test_ugrid_load.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        file_path = tests.get_data_path(
+            ("NetCDF", "unstructured_grid", "theta_nodal_xios.nc")
+        )
+
+        # cube = iris.load_cube(file_path, "theta")
+        # Note: cannot use iris.load, as merge does not yet preserve
+        # the cube 'ugrid' properties.
+
+        # Here's a thing that at least works.
+        loaded_cubes = CubeList(load_cubes(file_path, temp_xios_fix=True))
```
Comment (pp-mo) [inline iris_ugrid/tests/integration/test_ugrid_load.py:36]: The `temp_xios_fix` extension keyword needs to be removed here, as we're not using this now.
Cf #5 and https://github.com/SciTools-incubator/iris-ugrid/pull/4#discussion_r463710051

**Comment by trexfeathers** [comment]
### Regarding the possible restructure to `/lib/iris_ugrid/`:

I had interpreted this structure in SciTools libraries (e.g. [iris](https://github.com/SciTools/iris) or [cartopy](https://github.com/SciTools/cartopy)) as being 'behind the curve'. Having read @pp-mo's [above comment](https://github.com/SciTools-incubator/iris-ugrid/pull/4#issuecomment-668506361), and discussion at the Python Packaging Authority (see pypa/sampleproject#104), it now seems to me that using `/lib` or similar is actually 'ahead of the curve'! So I'm in favour of switching.

**Comment by trexfeathers** [comment]
Replaced by #24 (this had gone rather stale).

**Pull request closed without merging**